### PR TITLE
Model 2 Maternal Disorders Implementation and supporting Artifact

### PR DESCRIPTION
**Pull request by mattkappel**
Model 2 Implementation and supporting Artifact

### Description

- *Category*: data artifact, implementation
- *JIRA issue*: [MIC-2812](https://jira.ihme.washington.edu/browse/MIC-2812), [MIC-2808](https://jira.ihme.washington.edu/browse/MIC-2808)
- *Research reference*: [Maternal disorders](https://vivarium-research.readthedocs.io/en/latest/gbd2019_models/causes/maternal_disorders/index.html#cause-maternal-disorders)

Changes:
- Added maternal disorders cause of death
- Added maternal disorders cause of disability (final yld handling TBD)
- Added a prepostpartum state in between pregnant and postpartum states
- Refactored use of indices to use masks for ease of debugging in `on_time_step`.

### Verification and Testing
Able to `simulate run` model successfully.

### CODE REVIEW

File: src/vivarium_gates_iv_iron/components/pregnancy.py [hunk 1]
```diff
@@ -40,25 +45,54 @@ def setup(self, builder: Builder):
...
 
         outcome_probabilities = self.load_pregnancy_outcome_probabilities(builder)
         self.outcome_probabilities = builder.lookup.build_table(outcome_probabilities,
                                                                 key_columns=['sex'],
                                                                 parameter_columns=['age', 'year'])
-        # TODO remove age and sex columns when done debugging
-        self.population_view = builder.population.get_view(self.columns_created + ['age', 'sex'])
+
+        life_expectancy_data = builder.data.load("population.theoretical_minimum_risk_life_expectancy")
+        self.life_expectancy = builder.lookup.build_table(life_expectancy_data, parameter_columns=['age'])
```
Comment (rmudambi) [inline src/vivarium_gates_iv_iron/components/pregnancy.py:58]: This isn't year-specific data?

File: src/vivarium_gates_iv_iron/components/pregnancy.py [hunk 2]
```diff
@@ -40,25 +45,54 @@ def setup(self, builder: Builder):
...
+                                                                    parameter_columns=['age', 'year'])
+        maternal_disorder_incidence = builder.data.load("cause.maternal_disorders.incidence_rate").set_index([col for col in metadata.ARTIFACT_INDEX_COLUMNS if col != "location"])
+
+        self.probability_maternal_deaths = builder.lookup.build_table((maternal_disorder_csmr / conception_rate_data).reset_index(),
+                                                                      key_columns=['sex'],
+                                                                      parameter_columns=['age', 'year'])
+        self.probability_non_fatal_maternal_disorder = builder.lookup.build_table(((maternal_disorder_incidence - maternal_disorder_csmr) / conception_rate_data).reset_index(),
+                                                                      key_columns=['sex'],
+                                                                      parameter_columns=['age', 'year'])
+
```
Comment (rmudambi) [inline src/vivarium_gates_iv_iron/components/pregnancy.py:80]: This would be more legible if you could make these lines <= 100 characters long.

**Review by albrja — APPROVED** [review]

File: src/vivarium_gates_iv_iron/components/pregnancy.py [hunk 3]
```diff
@@ -40,25 +45,54 @@ def setup(self, builder: Builder):
...
+        self.life_expectancy = builder.lookup.build_table(life_expectancy_data, parameter_columns=['age'])
+
+        all_cause_mortality_data = builder.data.load("cause.all_causes.cause_specific_mortality_rate").set_index([col for col in metadata.ARTIFACT_INDEX_COLUMNS if col != "location"])
+        maternal_disorder_csmr = builder.data.load("cause.maternal_disorders.cause_specific_mortality_rate").set_index([col for col in metadata.ARTIFACT_INDEX_COLUMNS if col != "location"])
+        self.background_mortality_rate = builder.lookup.build_table((all_cause_mortality_data - maternal_disorder_csmr).reset_index(),
+                                                                    key_columns=['sex'],
+                                                                    parameter_columns=['age', 'year'])
+        maternal_disorder_incidence = builder.data.load("cause.maternal_disorders.incidence_rate").set_index([col for col in metadata.ARTIFACT_INDEX_COLUMNS if col != "location"])
+
+        self.probability_maternal_deaths = builder.lookup.build_table((maternal_disorder_csmr / conception_rate_data).reset_index(),
```
Comment (rmudambi) [inline src/vivarium_gates_iv_iron/components/pregnancy.py]: I'm assuming probability refers to the probability of death due to maternal disorders given that you are pregnant. If so, this equation is correct if `conception_rate = conceptions / person_time_total`. Is that what this data represents? It seems to me that a better definition of conception rate would be `conception_rate = conceptions / person_time_not_pregnant`

File: src/vivarium_gates_iv_iron/components/pregnancy.py [hunk 4]
```diff
@@ -98,65 +132,111 @@ def on_initialize_simulants(self, pop_data: SimulantData) -> None:
...
+        pop.loc[pregnant_this_step, "pregnancy_status"] = models.PREGNANT_STATE
+        pop.loc[pregnant_this_step, "pregnancy_outcome"] = pregnancy_outcome.loc[pregnant_this_step]
+        pop.loc[pregnant_this_step, "sex_of_child"] = sex_of_child.loc[pregnant_this_step]
+        pop.loc[pregnant_this_step, "birth_weight"] = birth_weight.loc[pregnant_this_step]
+        pop.loc[pregnant_this_step, "pregnancy_duration"] = pregnancy_duration
+        pop.loc[pregnant_this_step, "pregnancy_state_change_date"] = event.time
+
+        # Pregnancy to maternal disorder state and no maternal disorder state
+        maternal_disorder_this_step = maternal_disorder_this_step & pregnancy_ends_this_step
+        no_maternal_disorder_this_step = pregnancy_ends_this_step & ~maternal_disorder_this_step
```
Comment (rmudambi) [inline src/vivarium_gates_iv_iron/components/pregnancy.py]: It's weird that you flipped the order of these.

File: src/vivarium_gates_iv_iron/components/pregnancy.py [hunk 5]
```diff
@@ -171,16 +251,21 @@ def on_simulation_end(self):
...
                                .set_index(index_cols))
         postpartum_prevalence = (builder.data.load(data_keys.PREGNANCY.POSTPARTUM_PREVALENCE)
-                               .fillna(0)
-                               .set_index(index_cols))
-        # order of prevalences must match order of PREGNANCY_STATUSES
-        prevalences = pd.concat([not_pregnant_prevalence, pregnant_prevalence, postpartum_prevalence], axis=1)
+                                 .fillna(0)
+                                 .set_index(index_cols))
+        maternal_disorder_prevalence = pd.Series(0., index=postpartum_prevalence.index, name=models.MATERNAL_DISORDER_STATE)
+        no_maternal_disorder_prevalence = 1/6 * postpartum_prevalence
```
Comment (rmudambi) [inline src/vivarium_gates_iv_iron/components/pregnancy.py]: These `1/6` and `5/6` values should probably be defined in the `data_values` file.

File: src/vivarium_gates_iv_iron/constants/data_keys.py [hunk 6]
```diff
@@ -84,6 +84,51 @@ def log_name(self):
...
+
+    @property
+    def log_name(self):
+        return "maternal_disorders"
+
+
+MATERNAL_DISORDERS = __MaternalDisorders()
+
+
+class __MaternalHemorrhage(NamedTuple):
```
Comment (rmudambi) [inline src/vivarium_gates_iv_iron/constants/data_keys.py:111]: This is not used anywhere yet, correct?

File: src/vivarium_gates_iv_iron/components/pregnancy.py [hunk 7]
```diff
@@ -40,25 +45,54 @@ def setup(self, builder: Builder):
...
+        self.background_mortality_rate = builder.lookup.build_table((all_cause_mortality_data - maternal_disorder_csmr).reset_index(),
+                                                                    key_columns=['sex'],
+                                                                    parameter_columns=['age', 'year'])
+        maternal_disorder_incidence = builder.data.load("cause.maternal_disorders.incidence_rate").set_index([col for col in metadata.ARTIFACT_INDEX_COLUMNS if col != "location"])
+
+        self.probability_maternal_deaths = builder.lookup.build_table((maternal_disorder_csmr / conception_rate_data).reset_index(),
+                                                                      key_columns=['sex'],
+                                                                      parameter_columns=['age', 'year'])
+        self.probability_non_fatal_maternal_disorder = builder.lookup.build_table(((maternal_disorder_incidence - maternal_disorder_csmr) / conception_rate_data).reset_index(),
```
Comment (rmudambi) [inline src/vivarium_gates_iv_iron/components/pregnancy.py]: The same applies here. I'm guessing the denominator of `maternal_disorder_incidence` is total person time, in which case this equation works, but it seems like you really want the denominator to be something like person time susceptible to maternal disorders (which might be the same thing as person time not pregnant). Might be worth asking Ali about this.

**Review by rmudambi — APPROVED** [review]

File: src/vivarium_gates_iv_iron/constants/data_keys.py (hunk 6 above)
Comment (mattkappel) [reply src/vivarium_gates_iv_iron/constants/data_keys.py:111]: Correct.

File: src/vivarium_gates_iv_iron/components/pregnancy.py (hunk 3 above)
Comment (mattkappel) [reply src/vivarium_gates_iv_iron/components/pregnancy.py]: conception rate is the same as incidence of pregnancy, defined here: https://vivarium-research.readthedocs.io/en/latest/other_models/pregnancy/index.html#id7

File: src/vivarium_gates_iv_iron/components/pregnancy.py (hunk 7 above)
Comment (mattkappel) [reply src/vivarium_gates_iv_iron/components/pregnancy.py]: The code is following the model, but you're right it probably is worth asking. 
https://vivarium-research.readthedocs.io/en/latest/gbd2019_models/causes/maternal_disorders/index.html#id6

File: src/vivarium_gates_iv_iron/components/pregnancy.py (hunk 4 above)
Comment (mattkappel) [reply src/vivarium_gates_iv_iron/components/pregnancy.py]: Weird, but immaterial. There's a lot of cosmetic changes we ought to make.

File: src/vivarium_gates_iv_iron/components/pregnancy.py (hunk 5 above)
Comment (mattkappel) [reply src/vivarium_gates_iv_iron/components/pregnancy.py]: Agree

File: src/vivarium_gates_iv_iron/components/pregnancy.py (hunk 1 above)
Comment (mattkappel) [reply src/vivarium_gates_iv_iron/components/pregnancy.py:58]: Not sure.

**Pull request merged**